### PR TITLE
[core] Increment ref count when creating an ObjectRef to prevent object from going out of scope

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

When a Ray program first creates an ObjectRef (via ray.put or task call), we add it with a ref count of 0 in the C++ backend because the language frontend will increment the initial local ref once we return the allocated ObjectID, then delete the local ref once the ObjectRef goes out of scope. Thus, there is a brief window where the object ref will appear to be out of scope.

This can cause problems with async protocols that check whether the object is in scope or not, such as the previous bug fixed in #19910. Now that we plan to enable lineage reconstruction to automatically recover lost objects, this race condition can also be problematic because we use the ref count to decide whether an object needs to be recovered or not.

This PR avoids these race conditions by incrementing the local ref count in the C++ backend when executing ray.put() and task calls. The frontend is then responsible for skipping the initial local ref increment when creating the ObjectRef. This is the same fix used in #19910, but generalized to all initial ObjectRefs.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
I still need to fix up the ref counting in Java and C++ language frontends, but the Python version is ready for review.

**Comment by stephanie-wang** [comment]
Oh turns out the C++ frontend doesn't actually do ref counting.

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -361,10 +363,28 @@ cdef int prepare_actor_concurrency_groups(
...
         CoreWorker core_worker,
         Language language, args,
-        c_vector[unique_ptr[CTaskArg]] *args_vector, function_descriptor):
+        c_vector[unique_ptr[CTaskArg]] *args_vector, function_descriptor,
+        c_vector[CObjectID] *put_arg_ids):
+    try:
+        prepare_args_internal(core_worker, language, args, args_vector,
+                              function_descriptor, put_arg_ids)
+    except Exception as e:
+        # An error occurred during arg serialization. We must remove the
```
Comment (ericl) [inline python/ray/_raylet.pyx:375]: Hmm, is it possible the refs haven't been added yet?

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -1660,8 +1704,15 @@ cdef class CoreWorker:
                 args_vector,
                 CTaskOptions(
                     name, num_returns, c_resources, concurrency_group_name))
+            for put_arg_id in put_arg_ids:
```
Comment (ericl) [inline python/ray/_raylet.pyx:1707]: Comment why we're removing them?

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -1561,7 +1598,13 @@ cdef class CoreWorker:
                         is_asyncio or max_concurrency > 1,
                         max_pending_calls),
                     extension_data,
-                    &c_actor_id))
+                    &c_actor_id)
+
+            for put_arg_id in put_arg_ids:
+                CCoreWorkerProcess.GetCoreWorker().RemoveLocalReference(
+                    put_arg_id)
```
Comment (ericl) [inline python/ray/_raylet.pyx:1605]: Comment why we're removing them?

**Review by ericl — COMMENTED** [review]
Looks good, but I don't have the state to give a really detailed review. Maybe trigger nightly tests to flush out any edge cases?

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/_raylet.pyx:375]: We don't add to this list unless we've already incremented the local ref. Updated the variable name to clarify.

**Comment by stephanie-wang** [comment]
Ran nightly, I think all of these test failures are from master:

https://buildkite.com/ray-project/periodic-ci/builds/2519

**Comment by ericl** [comment]
Failing tests

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
LINT errors, etc. Can you rebase?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by stephanie-wang** [comment]
Test failures look unrelated (on the last commit it passed all but Java, and Java's now passing).

**Pull request merged**